### PR TITLE
[dtensor] implement scaled dot product attention (flash-attention)

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120298
* #120297

as titled, this PR implements the sdpa flash attention op in DTensor

Adding flash attention first but efficient attention and other attention
ops should be similar

fixes https://github.com/pytorch/pytorch/issues/120333

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

File: test/distributed/_tensor/test_matrix_ops.py [hunk 1]
```diff
@@ -264,6 +265,46 @@ def test_placement_comb(
...
+
+        from torch.nn.attention import sdpa_kernel, SDPBackend
+
+        with sdpa_kernel(backends=[SDPBackend.FLASH_ATTENTION]):
+            out = F.scaled_dot_product_attention(
+                query, key, value, dropout_p=0.0, is_causal=True
+            )
+            dist_out = F.scaled_dot_product_attention(
+                dist_query, dist_key, dist_value, dropout_p=0.0, is_causal=True
+            )
```
Comment (tianyu-l) [inline test/distributed/_tensor/test_matrix_ops.py:304]: do we want to check output are the same?

File: torch/distributed/_tensor/ops/matrix_ops.py [hunk 2]
```diff
@@ -128,3 +141,86 @@ def bmm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
...
+        spec_list = []
+        for specs in zip(*strategy_comb):
+            spec_list.append(DTensorSpec(mesh, tuple(specs)))
+
+        assert len(spec_list) == 6
+        input_expected_specs = spec_list[3:]
+        output_specs: List[Optional[DTensorSpec]] = list(spec_list[:3])
+        # fix up ouput_specs and fill in None for the int and empty tensor return values
+        for i in range(2, 8):
+            output_specs.insert(i, None)
```
Comment (tianyu-l) [inline torch/distributed/_tensor/ops/matrix_ops.py:206]: Would you please share a pointer to the specification (input, output) of the flash attention op? I only found [this](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/FlashAttentionKernel.cpp#L146) which it doesn't seem to match.

**Review by tianyu-l — COMMENTED** [review]
I wonder before this PR, what was the status of dtensor support for `scaled_dot_product_attention`? IIUC, in test_tp_examples.py we have a small transformer which already has F.sdpa computation compatible with dtensor.

**Comment by wanchaol** [comment]
> before this PR, what was the status of dtensor support for scaled_dot_product_attention? IIUC, in test_tp_examples.py we have a small transformer which already has F.sdpa computation compatible with dtensor.

@tianyu-l We didn't support this op at all before. Our tp_examples using TP APIs, i.e. ColwiseParallel and RowwiseParallel, where in the attention layer after ColwiseParallel, it would automatically calls `to_local()` to convert to torch.Tensor, so scaled_dot_product_attention was not using DTensor

File: torch/distributed/_tensor/ops/matrix_ops.py (hunk 2 above)
Comment (wanchaol) [reply torch/distributed/_tensor/ops/matrix_ops.py:206]: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L14526

File: test/distributed/_tensor/test_matrix_ops.py (hunk 1 above)
Comment (wanchaol) [reply test/distributed/_tensor/test_matrix_ops.py]: added

**Review by XilunWu — APPROVED** [review]
lgtm. Left a question about RNG.

File: test/distributed/_tensor/test_matrix_ops.py [hunk 3]
```diff
@@ -264,6 +265,61 @@ def test_placement_comb(
...
+            dist_out = F.scaled_dot_product_attention(
+                dist_query,
+                dist_key,
+                dist_value,
+                dropout_p=dropout_p,
+                is_causal=is_causal,
+            )
+            full_out = dist_out.full_tensor()
+            if self.rank == 0:
+                self.assertEqual(full_out, out)
```
[the remark this answers is not in the document]
Comment (wanchaol) [reply test/distributed/_tensor/test_matrix_ops.py]: It was mainly my debugging and forgot to toggle it, we can do a non-rank sensitive assertequal for this case, updated

**Pull request closed without merging**